delta-lake/src/main/java/org/apache/iceberg/delta/BaseSnapshotDeltaLakeTableAction.java

```diff
@@ -0,0 +1,340 @@
+   * @param deltaTableLocation the delta lake table's path
+   * @param newTableIdentifier the identifier of the new iceberg table
+   * @param deltaLakeConfiguration the hadoop configuration to access the delta lake table
+   */
+  public BaseSnapshotDeltaLakeTableAction(
+      Catalog icebergCatalog,
+      String deltaTableLocation,
+      TableIdentifier newTableIdentifier,
+      Configuration deltaLakeConfiguration) {
+    this.icebergCatalog = icebergCatalog;
+    this.deltaTableLocation = deltaTableLocation;
+    this.newTableIdentifier = newTableIdentifier;
+    this.newTableLocation = deltaTableLocation;
+    this.deltaLog = DeltaLog.forTable(deltaLakeConfiguration, deltaTableLocation);
+    this.deltaLakeFileIO = new HadoopFileIO(deltaLakeConfiguration);
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperties(Map<String, String> properties) {
+    additionalPropertiesBuilder.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableProperty(String name, String value) {
+    additionalPropertiesBuilder.put(name, value);
+    return this;
+  }
+
+  @Override
+  public SnapshotDeltaLakeTable tableLocation(String location) {
+    this.newTableLocation = location;
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    io.delta.standalone.Snapshot updatedSnapshot = deltaLog.update();
+    Schema schema = convertDeltaLakeSchema(updatedSnapshot.getMetadata().getSchema());
+    PartitionSpec partitionSpec = getPartitionSpecFromDeltaSnapshot(schema);
+    Transaction icebergTransaction =
+        icebergCatalog.newCreateTableTransaction(
+            newTableIdentifier,
+            schema,
+            partitionSpec,
+            newTableLocation,
+            destTableProperties(updatedSnapshot, deltaTableLocation));
+
+    long totalDataFiles = copyFromDeltaLakeToIceberg(icebergTransaction);
+    icebergTransaction.commitTransaction();
+    LOG.info(
+        "Successfully loaded Iceberg metadata for {} files in {}",
+        totalDataFiles,
+        deltaTableLocation);
+    return new BaseSnapshotDeltaLakeTableActionResult(totalDataFiles);
+  }
+
+  private Schema convertDeltaLakeSchema(io.delta.standalone.types.StructType deltaSchema) {
+    Type converted =
+        DeltaLakeDataTypeVisitor.visit(deltaSchema, new DeltaLakeTypeToType(deltaSchema));
+    return new Schema(converted.asNestedType().asStructType().fields());
+  }
+
+  private PartitionSpec getPartitionSpecFromDeltaSnapshot(Schema schema) {
+    List<String> partitionNames = deltaLog.snapshot().getMetadata().getPartitionColumns();
+    if (partitionNames.isEmpty()) {
+      return PartitionSpec.unpartitioned();
+    }
+
+    PartitionSpec.Builder builder = PartitionSpec.builderFor(schema);
+    for (String partitionName : partitionNames) {
+      builder.identity(partitionName);
+    }
+    return builder.build();
+  }
+
+  private long copyFromDeltaLakeToIceberg(Transaction transaction) {
```

Comment:
nit: do we still need this private method? seems like we can just move it back to the parent method as it's only called once and not a lot of logic.